compressors/shannon_coder.py

```diff
@@ -0,0 +1,156 @@
+"""Shannon Coding
+Shannon coding is a basic prefix free code. There is some confusion in literature about Shannon, Fano and
+Shannon-Fano codes, e.g. see Naming section of https://en.wikipedia.org/wiki/Shannon–Fano_coding.
+
+This document uses cumulative probability based method for shannon coding as described in the wiki article above.
+"""
+from typing import Any, Tuple
+from utils.bitarray_utils import float_to_bitarrays, BitArray
+from utils.test_utils import get_random_data_block, try_lossless_compression
+from compressors.prefix_free_compressors import PrefixFreeEncoder, PrefixFreeDecoder, PrefixFreeTree, BinaryNode
+from core.prob_dist import ProbabilityDist
+import math
+
+
+class ShannonTree(PrefixFreeTree):
+    """
+    Generates Codewords based on Shannon Coding. Allows generating PrefixFreeTree from the obtained codewords.
+    """
+
+    def __init__(self, prob_dist: ProbabilityDist):
+        self.prob_dist = prob_dist
+        # sort the probability distribution in decreasing probability and get cumulative probability which will be
+        # used for encoding
+        self.sorted_prob_dist = ProbabilityDist.get_sorted_prob_dist(prob_dist.prob_dict)
+        self.cum_prob_dict = self.sorted_prob_dist.cumulative_prob_dict
+        # construct the tree and set the root_node of PrefixFreeTree base class
+        super().__init__(root_node=self.build_shannon_tree())
+
+    def encode_alphabet(self, symbol) -> BitArray:
+        # compute the mid-point corresponding to the range of the given symbol
+        cum_prob = self.cum_prob_dict[symbol]
+
+        # compute encode length
+        encode_len = math.ceil(self.prob_dist.log_probability(symbol))
+
+        # the encode is the binary representation of the cumulative probability in ascending order,
+        _, code = float_to_bitarrays(cum_prob, encode_len)
+        return code
+
+    @staticmethod
+    def _build_tree_from_code(symbol, code, root_node) -> BinaryNode:
```

Comment:
I think this should sit outside Shannon code, maybe in the Prefix Tree code. Also I feel a better interface is to have:
```
    def build_shannon_tree(self):
        """
        For all symbols in the alphabet, get its code, and add it to the PrefixFreeTree.
        """
        root_node = BinaryNode(id=None)
        code = self.generate_codewords(s) # this internally uses encode_alphabet in a loop, returns dict from symbol to codeword
        root_node = self._build_tree_from_code(code)
        return root_node
```
where the `_build_tree_from_code` should take the entire code as a dict and generate the whole tree at once.

Reply:
Also wonder if it's natural to detect non-prefix codes as part of the `_build_tree_from_code` function

Reply:
thanks for the suggestion. Updated. 
Not sure about detecting comment: ideally there should be a detection during the creation of PrefixFreeClass and if at any point it becomes non-prefix free it should throw an error?